src/main/java/io/cdap/plugin/salesforce/SObjectsDescribeResult.java

```diff
@@ -65,26 +65,60 @@ public SObjectsDescribeResult(PartnerConnection connection, Collection<String> s
    *
    * @param connection Salesforce partner connection
-   * @param parentSObjects unique combinations of parent SObject names and their types
+   * @param name top level SObject name
+   * @param featuredSObjects unique combination of featured SObjects
    */
-  public SObjectsDescribeResult(PartnerConnection connection,
-                                Set<Map<String, SObjectDescriptor.SObjectType>> parentSObjects) {
+  public SObjectsDescribeResult(PartnerConnection connection, String name,
+                                Collection<SObjectDescriptor.FeaturedSObjects> featuredSObjects) {
     Map<String, DescribeSObjectResult> cache = new HashMap<>();
-    parentSObjects
-      .forEach(names -> {
-        try {
-          addSObjectDescribe(connection, names, cache);
-        } catch (ConnectionException e) {
-          throw new RuntimeException(e);
+    try {
+      // describe top level SObject first and store it to be used during schema creation
+      DescribeSObjectResult topLevelDescribe = describe(connection, name, cache);
+      addSObjectDescribe(name, topLevelDescribe.getFields());
+
+      // describe each featured SObjects combination
+      for (SObjectDescriptor.FeaturedSObjects obj : featuredSObjects) {
+        // at the beginning last describe is top level SObject describe result
+        DescribeSObjectResult lastDescribe = topLevelDescribe;
+
+        /*
+          If featured SObjects have relationship, describe it first.
+          For the query `SELECT Name, (SELECT LastName, Owner.UserRole.Id FROM Contacts) FROM Account`
+          `Contacts` is relationship SObject.
+         */
+        if (obj.hasRelationship()) {
+          String relationshipName = getRelationshipName(obj.getRelationship(), lastDescribe);
+          if (relationshipName == null) {
+            throw new IllegalArgumentException("Unable to find relationship field name: " + name);
+          }
+          lastDescribe = describe(connection, name, cache);
         }
-      });
+
+        /*
+          If featured SObjects have references, describe describe them as well.
+          For the query `SELECT Name, (SELECT LastName, Owner.UserRole.Id FROM Contacts) FROM Account`
+          `Owner` and `UserRole` are reference SObjects.
+         */
+        if (obj.hasReferences()) {
+          for (String reference : Objects.requireNonNull(obj.getReferences())) {
+            String referenceName = getReferenceName(reference, lastDescribe);
+            if (referenceName == null) {
+              throw new IllegalArgumentException("Unable to find reference field name: " + name);
+            }
+            lastDescribe = describe(connection, referenceName, cache);
+          }
+        }
+
+        // store final describe result to be used during schema creation
+        addSObjectDescribe(obj.getFullName(name), lastDescribe.getFields());
+      }
+    } catch (ConnectionException e) {
+      throw new RuntimeException(e);
```

Comment:
In general, we want to avoid catching a checked exception and throwing a runtime exception. This type of pattern makes it very difficult to do proper error handling, especially if it exists in multiple layers in the code. The caller might not be able to handle it in any meaningful way and may need to propagate it up, but that should be a decision that the caller makes and not this class.

In general we would also like to avoid constructors that throw exceptions or do work, such as making remote calls. A better pattern is to have a constructor that just takes the objectToFieldMap, and turn this into a static method that returns a SObjectsDescribeResult and throws a ConnectionException.

Reply:
Replaced with static methods calls which throws ConnectionException.